### PR TITLE
Make --use-bootstrap-image work in containers

### PR DESCRIPTION
**Pull request by praiskup**
Fixes: #691

### CODE REVIEW

**Comment by praiskup** [comment]
Mock prints with `--use-bootstrap-image --config-opts=package_manager=dnf5` something like this:
```
INFO: Using bootstrap image: registry.fedoraproject.org/fedora:rawhide
INFO: Pulling image: registry.fedoraproject.org/fedora:rawhide
...
INFO: Copy content of container registry.fedoraproject.org/fedora:rawhide to /var/lib/mock/fedora-rawhide-x86_64-bootstrap/root
INFO: mounting registry.fedoraproject.org/fedora:rawhide with podman image mount
INFO: image registry.fedoraproject.org/fedora:rawhide as /var/lib/containers/storage/overlay/7821387bce6888d9804d3a725dde054835057205456e45466ef9e6c5b873d62c/merged
INFO: umounting image registry.fedoraproject.org/fedora:rawhide (/var/lib/containers/storage/overlay/7821387bce6888d9804d3a725dde054835057205456e45466ef9e6c5b873d62c/merged) with podman image umount
...
INFO: Using package manager from the bootstrap image
INFO: Searching for /bin/dnf5 in image
INFO: Searching for /bin/dnf in image
INFO: Package manager dnf detected and used
Start(bootstrap): installing dnf5 tooling
...
Installing:
 dnf5                                                        x86_64                                                5.0.9-3.fc39                                                 fedora                                                438 k
 dnf5-plugins                                                x86_64                                                5.0.9-3.fc39                                                 fedora                                                 80 k
...
INFO: Using package manager from the bootstrap image
INFO: Searching for /bin/dnf5 in image
INFO: Switching package manager from dnf to the dnf5
... from now on, DNF5 in bootstrap is used to install buildroot ...
Start: chroot init
...
INFO: Package manager dnf5 detected and used
Start: installing minimal buildroot with dnf5
...
```

Tested inside a RHEL-8-based container.

**Comment by praiskup** [comment]
@xsuchy can you please take a look at this one?

File: mock/py/mockbuild/package_manager.py [hunk 1]
```diff
@@ -91,12 +91,27 @@ def package_manager_class_fallback(desired, config_opts, bootstrap):
     raise Exception("No package from {} found".format(fallbacks[desired]))
 
 
+def package_manager_class_fallback_from_image(desired, bootstrap):
+    if desired not in fallbacks:
+        raise RuntimeError(f'Unexpected package manager "{desired}"')
+    path = bootstrap.make_chroot_path()
+    for manager in fallbacks[desired]:
+        getLog().info("Searching for /bin/%s in image", manager)
+        if os.path.exists(os.path.join(path, "usr", "bin", manager)):
```
Comment (xsuchy) [inline mock/py/mockbuild/package_manager.py:100]: /bin in Log() vs /usr/bin in os.path

File: mock/py/mockbuild/buildroot.py [hunk 2]
```diff
@@ -140,7 +160,7 @@ def initialize(self, prebuild=False, do_log=True):
             self._lock_buildroot(exclusive=True)
             self._init(prebuild=prebuild, do_log=do_log)
         except BuildRootLocked:
-            pass
+            self._init_locked()
```
Comment (praiskup) [inline mock/py/mockbuild/buildroot.py:163]: This was a missing bit for making #627 to work (22c8fdcbd0f1de50b942a5792a0fa6e88c620946 is the first needed patch).

File: mock/py/mockbuild/podman.py [hunk 3]
```diff
@@ -52,35 +35,37 @@ def install_pkgmgmt_packages(self):
...
-        podman = subprocess.Popen(cmd_podman, stdout=subprocess.PIPE)
-        cache_file = open(cache_file_name, "w")
-        cmd_compressor = [compress_program, "--stdout"]
-        compressor = subprocess.Popen(cmd_compressor, stdin=podman.stdout, stdout=cache_file)
-        compressor.communicate()
-        podman.communicate()
-        cache_file.close()
+    def exec(self, command):
+        """ exec command in container """
+        util.do(command, chrootPath=self.buildroot.make_chroot_path(),
```
Comment (xsuchy) [inline mock/py/mockbuild/podman.py:40]: Either the docstring is false or the body of this function. I fail to perceive how this exec command in container.

File: mock/py/mockbuild/podman.py [hunk 4]
```diff
@@ -52,35 +35,37 @@ def install_pkgmgmt_packages(self):
...
-        podman.communicate()
-
-    @traceLog()
-    def remove(self):
-        """ remove the container """
-        cmd = ["podman", "rm", "-f", self.container_id]
-        util.do(cmd)
-        self.container_id = None
+        with self.mounted_image() as mount_path:
+            cmd_podman = [tar_cmd, "-C", mount_path, "-c", "."]
```
Comment (xsuchy) [inline mock/py/mockbuild/podman.py:63]: I would put a comment above this line. When you replaced "podman export" with tar_cmd it is less obvious what you are doing here. 
```
# create a tarball from mounted podman image and send it to PIPE
```

File: mock/py/mockbuild/package_manager.py [hunk 5]
```diff
@@ -91,12 +91,29 @@ def package_manager_class_fallback(desired, config_opts, bootstrap):
     raise Exception("No package from {} found".format(fallbacks[desired]))
 
 
+def package_manager_class_fallback_from_image(desired, bootstrap):
+    if desired not in fallbacks:
+        raise RuntimeError(f'Unexpected package manager "{desired}"')
+    path = bootstrap.make_chroot_path()
+    for manager in fallbacks[desired]:
+        binary = os.path.join(path, "usr", "bin", manager)
```
Comment (xsuchy) [inline mock/py/mockbuild/package_manager.py:99]: ??? Why not self.config[f'{pm}_command'] ?  I know that we have no config nor self here, but that can be done.

**Comment by xsuchy** [comment]
Otherwise ook.
This was really hard to review :)

File: mock/py/mockbuild/package_manager.py (hunk 5 above)
Comment (praiskup) [reply mock/py/mockbuild/package_manager.py:99]: This is the `package_manager_class_fallback_from_image()` method - we try to find a pre-existing package manager (any) to install the target package manager.  This is to mimic what we do *on host* when installing bootstrap.  I'll dump there a docstring.

**Comment by praiskup** [comment]
> This was really hard to review :)

/me nods, thank you very much for looking!

File: mock/py/mockbuild/podman.py (hunk 3 above)
Comment (xsuchy) [reply mock/py/mockbuild/podman.py:42]: Hmm, the docstring is good now, but... I thought that point of Podman() class is to execute inside container using a podman. But this change the logic of Podman to execute the command using nspawn inside of the directory that was extracted from podman image.

Comment (praiskup) [reply mock/py/mockbuild/podman.py:42]: Nah, I missed this comment - that is right, and actually the point of this change. As long as we support Mock-in-containers/toolbox, it is not yet an easy thing to unconditionally spawn (nested, either Podman or Nspawn) containers by Mock.  It is though OK to let the automatic image download and mount logic on Podman (works fine in any rootless --privileged container).

**Pull request merged**